src/validation.cpp

```diff
@@ -909,12 +909,14 @@ bool MemPoolAccept::ReplacementChecks(Workspace& ws)
     TxValidationState& state = ws.m_state;
 
     CFeeRate newFeeRate(ws.m_modified_fees, ws.m_vsize);
-    // It's possible that the replacement pays more fees than its direct conflicts but not more
-    // than all conflicts (i.e. the direct conflicts have high-fee descendants). However, if the
-    // replacement doesn't pay more fees than its direct conflicts, then we can be sure it's not
-    // more economically rational to mine. Before we go digging through the mempool for all
-    // transactions that would need to be removed (direct conflicts and all descendants), check
-    // that the replacement transaction pays more than its direct conflicts.
+    // The replacement transaction must have a higher feerate than its direct conflicts. Before
+    // ancestor packages were used in mining, this was equivalent to checking that our mining code
+    // would prefer the replacement. Now, it serves as a fail-fast mechanism.
```

Comment:
> Now, it serves as a fail-fast mechanism.

This sentence strikes me as odd; I think it gives the idea that perhaps this check is redundant (but faster to calculate) with the check that involves all descendants, which isn't the case.

If it were me, I'd add a comment like this to the code:

```
The replacement transaction must have a higher feerate than its direct conflicts.
- The motivation for this check is to ensure that the replacement transaction is preferable for block-inclusion, compared to what would be removed from the mempool.
- This logic predates ancestor feerate-based transaction selection, which is why it doesn't consider feerates of descendants.
- Ancestor feerate-based transaction selection has made this comparison insufficient to guarantee that this is incentive-compatible for miners, because it is possible for a descendant transaction of a direct conflict to pay a higher feerate than the transaction that might replace them, under these rules.
```

Reply:
Thanks for writing an alternative! I've changed the comment to be this.